WIP: Attempt of doing threading in the routing part of the algorithm: no luck

mauro3 opened the pull request:
For some reason the threaded version is much slower than the non-threaded.
This might be because of memory issues and cache threashing.  However, even when run
serially, the `@threads` decorated loop is much much slower???

If someone has any ideas, let me know.

visr commented:
I haven't worked with threading that much yet, but I tried to give this branch a spin on julia nightly. It was giving intermittent crashes that went away if I removed `@threads`, so I think before looking at performance we should look into correctness / race conditions a bit more.

Specifically this line:
https://github.com/mauro3/WhereTheWaterFlows.jl/blob/bf98aba4d5f95041dcceeff581cc78902352929a/src/WhereTheWaterFlows.jl#L347
causes intermittent "UndefRefError: access to undefined reference" for me.

Also looks like `Statistics` needs to be added to the deps in the .toml still.

mauro3 commented:
Thanks for looking at this.

Yes, that line sure was wrong, tnx!  Did not cause any issues with my runs though (surprising, considering I ran it over pretty large DEMs).  It was also wrong on master BTW, now fixed.

Did other `@threads` give you problems too?  The general strategy is that looping over catchments (aka colors, pits) should be thread-safe as different catchment do not interact.  So the other ones should be ok:
- https://github.com/mauro3/WhereTheWaterFlows.jl/blob/bf98aba4d5f95041dcceeff581cc78902352929a/src/WhereTheWaterFlows.jl#L282
- https://github.com/mauro3/WhereTheWaterFlows.jl/blob/bf98aba4d5f95041dcceeff581cc78902352929a/src/WhereTheWaterFlows.jl#L224

The `Statistics` I just used from some `@show` debugging, now removed.

OT: what is the size of DEMs you'd be running such a code over?  I'm currently looking at Antarctica with a DEM of up to 13000^2 points (2e8).  Single core performance is not so bad, about 15min, but a bit faster would be nice.  For that DEM most of the time (>50%) is spend in `flowrouting`, followed by `drainpits`.

mauro3 commented:
Superseded by #12

visr commented:
> Did other @threads give you problems too?

Nope, right now on master running tests repeatedly I never see an issue.

> OT: what is the size of DEMs you'd be running such a code over?

mauro3 commented:
> Nice job on #12! It's a good sign if a PR moves a bottleneck to a different function :)

I guess now I should re-open #6 ;-)

> Varies a lot I think, often 1e6 - 1e8.

Let me know how that goes.  I think, if you don't have  more than several 100k pits, then this package should work for you.

visr commented:
Will do! At the moment I'm not using it. I am working on https://github.com/Deltares/Wflow.jl, but that is just the hydrological model, preprocessing is still done in Python at the moment.

mauro3 commented:
Sounds good.  @luraess and I are working on a diffusion wave code at the moment (on the GPU), might be of interest to you too: https://github.com/mauro3/WhereTheWaterAlsoFlows.jl

visr commented:
Oh that is definitely interesting! We also were planning to add the diffusive wave to Wflow.jl.

mauro3 commented:
Cool.  I'll let you know if we get to anything usable. We only looked at steady states so far, but presumably you're interested in the transient case?
(Note, that you can't use our package yet as one of the deps is not open-sourced yet)

visr commented:
Thanks! Yes I'd also be interested in the transient case.
